---
title: "Interaction"
output: html_document
runtime: shiny
---

```{r, echo = FALSE}

  library(shiny)  
  library(DESeq2)
  library(stringr)
  library(dplyr)
  library(ggplot2)
  library(reshape2)
  library(plotly)
  library(biomaRt)
  source("helper_functions.r")

  
  counts <- read.table("htseq_counts_matrix.txt", sep="\t", header=T, row.names=1)
  coldata <- read.table("metadata_wo_outlier.txt", sep="\t", header=T, row.names=1)
  
  coldata$Genotype <- str_replace(coldata$Genotype, "RGS20_", "") 
  counts <- counts[rowSums(counts >= 1) >= 2, ]

  all(rownames(coldata) %in% colnames(counts))
  counts = counts[, rownames(coldata)]
  all(rownames(coldata) == colnames(counts))

  dds <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ Genotype + Treatment + Genotype:Treatment)

  dds$Treatment <- relevel(dds$Treatment, 'Naive')
  dds$Genotype <- relevel(dds$Genotype, 'WT')

  dds <- DESeq(dds)
  
  GetAnnotationFromBiomart <- function(species) {
  if (species == "hg38") {
    dataset="hsapiens_gene_ensembl"
  } else if (species == "mm10") {
    dataset="mmusculus_gene_ensembl"  
  } else if (species == "rn6") {
    dataset="rnorvegicus_gene_ensembl"
  }
  mart <- useMart(biomart="ENSEMBL_MART_ENSEMBL", dataset=dataset, host="aug2017.archive.ensembl.org")
  annotation_biomart <- getBM(attributes = c("ensembl_gene_id","external_gene_name","chromosome_name","start_position", 
                                             "end_position", "gene_biotype", "description"), 
                              filters = "ensembl_gene_id", values = rownames(dds), mart = mart) 
  rownames(annotation_biomart) <- annotation_biomart$ensembl_gene_id
  annotation_biomart <- subset(annotation_biomart, select=c(-`ensembl_gene_id`))
  return(annotation_biomart)
}

  AnnotateDifflist <- function(annotations, difflist) {
  annotations <- annotations[rownames(difflist), ]
  print(all(rownames(annotations) == rownames(difflist)))
  annot_difflist <- cbind(difflist, annotations)
  annot_difflist <- subset(annot_difflist, select=c(7,1,2,3,4,5,6,8,9,10,11,12))
  colnames(annot_difflist) <- c("gene_name", "baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue", 
                                "padj", "chromosome", "gene_start",
                                "gene_end", "gene_type", "description")
  return(annot_difflist)
  }
  
  ExtractDifferentialListsMultiFactor <- function(filename, var1, var2=NULL) {
  if (is.null(var2)) {
    difflist <- results(dds, name=var1)
    difflist <- AnnotateDifflist(annotation, difflist) 
    difflist <- difflist[order(difflist$pvalue), ]
    difflist$gene_id <- rownames(difflist)
    write.table(difflist, file = filename, sep = "\t", quote = F, row.names=F)
    return(difflist)
    
  } else {
    difflist <- results(dds, contrast = list(c(var1, var2)))
    difflist <- AnnotateDifflist(annotation, difflist) 
    difflist <- difflist[order(difflist$pvalue), ]
    difflist$gene_id <- rownames(difflist)
    write.table(difflist, file = filename, sep = "\t", quote = F, row.names=F)
    return(difflist)
  }
  }
  
  ObtainNormalizedCounts <- function() {
  data_normalized = counts(dds, normalized = T)
  return(data.frame(data_normalized))
  }
  
  annotation <- GetAnnotationFromBiomart("mm10")
  species="mm10"
  interaction <- ExtractDifferentialListsMultiFactor("interaction.txt", "GenotypeKO.TreatmentCFA") 
  interaction <- interaction[complete.cases(interaction), ] 
  CFA_vs_Naive_in_WT <- ExtractDifferentialListsMultiFactor("CFA_vs_Naive_in_WT.txt", "Treatment_CFA_vs_Naive") 
  CFA_vs_Naive_in_KO <- ExtractDifferentialListsMultiFactor("CFA_vs_Naive_in_KO.txt", "Treatment_CFA_vs_Naive", "GenotypeKO.TreatmentCFA") 
  norm_counts <- ObtainNormalizedCounts()
  
```




```{r, echo = FALSE}
UnderstandInteraction <- function(interaction_list, norm_counts, coldata) {

  shinyApp(
              ui <- fluidPage(
                
                fluidRow(
                  selectInput('gene_of_interest', 'Choose gene:', choices=interaction$gene_name,
                              multiple=TRUE, selected=c("Pde1a"))
                ),
                
                fluidRow(
                  plotOutput('ggplot_img', height = "450px"),
                ),
                
                fluidRow(
                  span(textOutput('padj'), style="font-size:16px; font-style: bold"),
                  span(textOutput('description'), style="font-size:16px; font-style: bold"))
              ),
    

    server <- function(input, output, session) {

                selectedData = reactive({
                  gene_ID <- row.names(interaction[interaction$gene_name %in% input$gene_of_interest, ])
                  exp_test_gene <- norm_counts[row.names(norm_counts) %in% gene_ID, ]
                  exp_test_gene <- t(as.data.frame(exp_test_gene))
                  if (all(row.names(coldata) == row.names(exp_test_gene))) {
                    exp_test_gene <- melt(cbind(exp_test_gene, coldata))
                  } else {
                    break
                  }
                  exp_test_gene
                  
                })
                
                gene_ids <- reactive({
                    dose.labs <- input$gene_of_interest
                    names(dose.labs) <- row.names(interaction[interaction$gene_name %in% input$gene_of_interest, ])
                    dose.labs
                  })
                
                output$ggplot_img <- renderPlot(
                  
                  ggplot(selectedData(), aes(x=Treatment, y=value, color=Treatment)) +
                    geom_point(size=3) + 
                    theme_linedraw(base_size = 18) +
                    ylab("Normalized Expression") +
                    facet_grid( variable ~ factor(Genotype, levels=c("WT", "KO")), labeller = labeller(variable = gene_ids())) +
                    expand_limits(y = 0) +
                    stat_summary(
                      fun = median,
                      geom = "line",
                      color="black",
                      aes(group=Genotype), position = position_dodge(width=0.9)
                    ),
                height = 450, res= 96 )
              
      
    }
  )
  
}
```


```{r}
interaction_widget <- UnderstandInteraction(interaction, norm_counts, coldata)
interaction_widget
```



